src/Data/Biapplicative.hs

```diff
@@ -84,6 +88,40 @@ biliftA3 :: Biapplicative w => (a -> b -> c -> d) -> (e -> f -> g -> h) -> w a e
 biliftA3 f g a b c = bimap f g <<$>> a <<*>> b <<*>> c
 {-# INLINE biliftA3 #-}
 
+data Mag a b t where
+  Pure :: t -> Mag a b t
+  Map :: (x -> t) -> Mag a b x -> Mag a b t
+  Ap :: Mag a b (t -> u) -> Mag a b t -> Mag a b u
+  One :: a -> Mag a b b
+
+-- | Traverse a 'Traversable' container in a 'Biapplicative'.
+traverse2 :: forall t a b c f. (Traversable t, Biapplicative f)
+          => (a -> f b c) -> t a -> f (t b) (t c)
+traverse2 f0 xs0 = go m m
+  where
+    m :: Mag a x (t x)
+    m = traverse One xs0
+
+    go :: forall x y. Mag a b x -> Mag a c y -> f x y
+    go (Pure t) (Pure u) = bipure t u
+    go (Map f x) (Map g y) = bimap f g (go x y)
+    go (Ap fs xs) (Ap gs ys) = go fs gs <<*>> go xs ys
+    go (One x) (One _) = f0 x
+    go _ _ = error "Impossible: the arguments are always the same."
```

Comment:
Perhaps you could get rid of this unreachable case by adding an extra type parameter to `Mag` and giving each constructor a unique tag?

Reply:
Do you have a way? I don't see a way, but am open to suggestions.

Reply:
Hm. Now that I look closer, this is much harder than I thought. My inclination was to augment `Mag` like so:

```haskell
-- Type-level tags.
data Pure
data Map z
data Ap z1 z2
data One

data Mag a b z t where
  Pure :: t -> Mag a b Pure t
  Map :: (x -> t) -> Mag a b z x -> Mag a b (Map z) t
  Ap :: Mag a b z1 (t -> u) -> Mag a b z2 t -> Mag a b (Ap z1 z2) u
  One :: a -> Mag a b One b
```

This jives well with `traverse2` itself:

```haskell
-- | Traverse a 'Traversable' container in a 'Biapplicative'.
traverse2 :: forall t a b c f. (Traversable t, Biapplicative f)
          => (a -> f b c) -> t a -> f (t b) (t c)
traverse2 f0 xs0 = go m m 
  where
    m :: Mag a x One (t x)
    m = traverse One xs0
    
    go :: forall x y z. Mag a b z x -> Mag a c z y -> f x y
    go (Pure t) (Pure u) = bipure t u
    go (Map f x) (Map g y) = bimap f g (go x y)
    go (Ap fs xs) (Ap gs ys) = go fs gs <<*>> go xs ys
    go (One x) (One _) = f0 x
```

But it does _not_ mesh particularly well with the `Functor` and `Applicative` instances you're trying to write. I don't have any concrete ideas for how to make that work, so perhaps what you have here will suffice.